[Validator] Improve Iban Validation

afurculita opened the pull request:
| Q | A |
| --- | --- |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | https://github.com/symfony/symfony/issues/13802 |
| License | MIT |
| Doc PR |  |

This PR adds 2 more checks for better IBAN validation: the validation of the country code (represented by the first 2 letters) and the format validation according to the country format regex based on http://www.swift.com/dsp/resources/documents/IBAN_Registry.pdf.

Commit:
Better Iban Validation

Added more country-based tests, added new error type

The error type `NOT_SUPPORTED_COUNTRY_CODE_ERROR` has been added for the
IBANs from unsupported countries, for which we don't have defined a
format. We will not check anymore if a country code against the Intl
component.
The tests have been completed with more tests for each contry we have
formats defined.

The  4 character length check and the case check has been removed. The message code constants

`TOO_SHORT_ERROR` and `INVALID_CASE_ERROR` has been deprecated

javiereguiluz commented:
@afurculita thanks for submitting this nice pull request.

I'd like to ask you something: reading the attached PDF, I found some constraints specific to some countries. For example: for Estonia (`EE`) *"first digit of bank code can not be 0"* (page 29 of the PDF). The proposed regular expressions don't take into account these details. I agree with that to not overcomplicate regular expressions, but I'd like to know what others think about this.

Another question is related to France (page 31). There are some *régions d'outre-mer* which use the same format but with a different country code (Guyana (GF), Guadeloupe (GP), Martinique (MQ), etc.) Should we take into account these details or again is too complicated and not that important?

stof commented on src/Symfony/Component/Validator/Constraints/IbanValidator.php:
```diff
@@ -26,6 +27,84 @@
+     *
+     * @var array
+     */
+    protected $countryFormats = array(
```
should be private

afurculita replied:
Changed it to private

stof commented on src/Symfony/Component/Validator/Constraints/IbanValidator.php:
```diff
@@ -64,52 +143,74 @@ public function validate($value, Constraint $constraint)
-                    ->addViolation();
+                              ->setParameter('{{ value }}', $this->formatValue($value))
+                              ->setCode(Iban::INVALID_CHARACTERS_ERROR)
+                              ->addViolation();
```
this must be reverted. Please don't change the Symfony coding standard. We use 4 spaces per indentation level

stof commented on src/Symfony/Component/Validator/Tests/Constraints/IbanValidatorTest.php:
```diff
@@ -11,12 +11,20 @@
 {
+    protected function setUp()
+    {
+        IntlTestHelper::requireFullIntl($this);
```
does it really require a full Intl ? I don't think it does

afurculita replied:
Changed it to `requireIntl`

SpacePossum commented:
Hi and thanks for the PR :)
I think is a great idea to improve this validator. Beside the questions @javiereguiluz raised I have one.
I did a little check against a project of mine running https://code.google.com/p/php-iban/source/browse/tags/1.4.7/registry.txt on top of/beside the Symfony validator and spotted two things. That project also doesn't echek the `EE` "not be 0". The second is about your changes;
what about the following country codes?
Aland Islands AX
United Arab Emirates AE
BL, GF, GP, MF, MQ, NC, PF, PM, QA, PM, RE, TF, YT, WF
I didn't check, but are these unofficial/canidates?

I think a great addition would be to have to the option to check entries against:
- official countries only
- official + candidates
- SEPA only
- custom set of countries codes only (maybe on top of the SEPA list)

Would such a change be possible?

afurculita commented:
@javiereguiluz @SpacePossum  first thank you for your attention on this PR.

> reading the attached PDF, I found some constraints specific to some countries. For example: for Estonia (EE) "first digit of bank code can not be 0" (page 29 of the PDF). The proposed regular expressions don't take into account these details.

I was thinking about that too. I've seen on [Wikipedia](http://www.wikiwand.com/en/International_Bank_Account_Number#/IBAN_formats_by_country), that there are more rules like the one you found. For example: Macedonia always has 07 as the IBAN check digits, same Mauritania with 13. It is very unlikely that these will change anytime soon, so I was thinking to add them too. What do you think?

> There are some régions d'outre-mer which use the same format but with a different country code (Guyana (GF), Guadeloupe (GP), Martinique (MQ), etc.) Should we take into account these details or again is too complicated and not that important?

We may add these too. They are having the same format as France (*'FR\d{2}\d{5}\d{5}[\dA-Z]{11}\d{2}'*), even if their ISO code is not *FR*. But they can be useful if we implement some country groups functionality in case someone wants to accept credit cards only from a determined list of countries. 

> what about the following country codes? Aland Islands AX, United Arab Emirates AE, BL, GF, GP, MF, MQ, NC, PF, PM, QA, PM, RE, TF, YT, WF. I didn't check, but are these unofficial/canidates?

I have included validation for AE. AX has the same format as Finland. French Guyana (GF), Guadeloupe (GP), Martinique (MQ), Reunion (RE), French Polynesia (PF), French Southern Territories (TF), Mayotte (YT), New Caledonia (NC), Saint Barthelemy (BL), Saint Martin (French part) (MF), Saint Pierre et Miquelon (PM), and Wallis and Futuna Islands (WF) has the same format as France. We can add them for the same reason I specified in the previous answer. 

> I think a great addition would be to have to the option to check entries against: official countries only, official + candidates, SEPA only, custom set of countries codes only (maybe on top of the SEPA list)

I was thinking about that too. I think it will be enough to add a way to specify a group of countries and have a predefined one: SEPA (or 2, by adding the ALL group).

What do you think, guys?

sstok commented on src/Symfony/Component/Validator/Constraints/IbanValidator.php:
```diff
@@ -64,51 +143,73 @@ public function validate($value, Constraint $constraint)
+
+        // ...and have a valid format, if available
+        if (isset($this->countryFormats[$countryCode]) &&
+            false == preg_match('/^'.$this->countryFormats[$countryCode].'$/', $canonicalized)
```
This look wrong, preg_match will return the number of matches and false on error. If there is an error its a bug.

afurculita replied:
So you want, in case `preg_match` will return `false`, to bypass this check? Or to return a constraint violation or exception?

sstok replied:
If preg_match returns false this is because because the regex is invalid in which case PHP will also give an E_WARNING (and thus the application crashes).

Symfony prefers to use strict comparisons, and as the warning is not suppressed, using a strict comparison is better then casting integer to boolean. :)

afurculita replied:
How about just using `!preg_match` instead of `false == preg_match`?

sstok replied:
That could work :)

SpacePossum commented:
@afurculita Thanks for the feedback. I think you are on the right track. Having a rule for each ISO code  might help to speed up checking against a configured list of ISO codes people want to accept. (and it makes it easy to maintain if such a code is changed, removed or a new code is added), so :+1:  :)

webmozart commented on src/Symfony/Component/Validator/Constraints/IbanValidator.php:
```diff
@@ -26,6 +27,84 @@
+     *
+     * @var array
+     */
+    private $countryFormats = array(
```
I think it's sufficient to name this just `$formats`.

webmozart commented on src/Symfony/Component/Validator/Constraints/IbanValidator.php:
```diff
@@ -64,51 +143,73 @@ public function validate($value, Constraint $constraint)
+        $countryCode = substr($canonicalized, 0, 2);
+        $countries = Intl::getRegionBundle()->getCountryNames();
+
+        if (!ctype_alpha($countryCode) || !isset($countries[$countryCode])) {
```
I think you should check `$this->formats[$countryCode]` here. From the perspective of the IbanValidator, I think it's safe to say that a country code is invalid if it is not acceptable for an IBAN.

webmozart commented on src/Symfony/Component/Validator/Tests/Constraints/IbanValidatorTest.php:
```diff
@@ -11,12 +11,20 @@
 {
+    protected function setUp()
+    {
+        IntlTestHelper::requireIntl($this);
```
This can be removed if the dependency on intl is removed.

webmozart commented on src/Symfony/Component/Validator/Tests/Constraints/IbanValidatorTest.php:
```diff
@@ -173,19 +181,25 @@ public function testInvalidIbans($iban, $code)
+            array('NO93 8601117 947', Iban::INVALID_FORMAT_ERROR),
+            array('CH930072011623852957', Iban::INVALID_FORMAT_ERROR),
+            array('CY170020 128 0000 0012 0052 7600', Iban::INVALID_FORMAT_ERROR),
+            array('FR76', Iban::INVALID_FORMAT_ERROR),
```
I would like to see tests for

* valid IBANs from every country
* invalid IBANs from every country

SpacePossum replied:
```php
            // BLZ violation, valid length and checksum. Known issue.
            // array('NL08 TEST 0300 0652 64'),

            // Country code violation, valid length and checksum
            array('XX10 RABO 0300 0652 64', Iban::INVALID_COUNTRY_CODE_ERROR),
            array('ZZ40OJIC391330359QFOEXXQX4G5B8JZLQ2', Iban::INVALID_COUNTRY_CODE_ERROR),
```

I would like these to be added, since these are now false positives.

SpacePossum replied:
> I would like to see tests for

> valid IBANs from every country

@afurculita please add Brazil to the valid list, then all countries you are checking for have a valid test

```php
            array('BR9700360305000010009795493P1'), //Brazil
```

> invalid IBANs from every country

Bucket list:
AD, AE, AL, AT, AZ, BA, BE, BG, BH, BR, CR, CZ, DE, DO, DK, EE, ES, FI, FO, GB, GE, GI, GL, 
GR, GT, HR, HU, IE, IL, IS, IT, KW, KZ, LB, LI, LT, LU, LV, MC, MD, ME, MK, MR, MT, MU, NL, 
PK, PL, PS, PT, RO, RS, SA, SE, SI, SK, SM, TN, TR, VG,

sstok replied:
NL (valid): NL43 INGB 0662352688

This is my old number, no longer in use but it should be usable.

webmozart commented:
Thank you for working on this @afurculita! I think we should limit the scope of this PR and work on formats only. Other features like country groups should be implemented in separate PRs.

afurculita commented:
Thanks @SpacePossum and @webmozart for the help with this PR.

I have committed new changes: https://github.com/afurculita/symfony/commit/d7533daf660ab60ebadc9c3de611204decc19274. I removed the dependency on Intl, added 20 more countries to our formats list, added a new error type and added a lot more tests. As @webmozart  recommended the other features like country groups it will be moved to other PRs.

SpacePossum commented:
Looking good, thanks!
Nice to see the dependency on Intl is removed. I agree with you and @webmozart to address other features in new PR's.

Failing tests are not related to the changes made in this PR and I don't know were fabbot went (having the weekend off maybe ;) )

I'm not sure if I scanned all the test correctly, but did you add a test for some IBAN's that have a valid country code and a valid checksum but are in an invalid format for that country?

afurculita commented:
I have added tests for IBANs with invalid formats and also for IBANs with valid formats but invalid checksum. If the format is invalid, the validator will be stopped before reaching the checksum validation. Should I still add tests for IBAN's that have a valid country code and a valid checksum but are in an invalid format for that country?

SpacePossum commented:
ah that makes sense, you are right, no point adding more tests :+1:

fabpot commented on src/Symfony/Component/Validator/Constraints/IbanValidator.php:
```diff
@@ -64,8 +176,44 @@ public function validate($value, Constraint $constraint)
+        }
+
+        // ...contain uppercase characters only
+        if ($canonicalized !== strtoupper($canonicalized)) {
```
Can't we just uppercase automatically. Rejecting an IBAN just because of something we can do automatically seems weird to me. Or am I missing something here?

SpacePossum replied:
it would change current behaviour;
https://github.com/symfony/symfony/blob/2.7/src/Symfony/Component/Validator/Constraints/IbanValidator.php#L102

Wouldn't this be a task for a transformer? I'm not sure but some people don't like validator to change the value its testing. I'm in favor of removing the case test btw.

fabpot commented:
fabbpot is back from holidays with some good suggestions :)

rgeraads commented:
I think this change would make the 4 character length check obsolete as well :)
https://github.com/afurculita/symfony/blob/better-iban-validation/src/Symfony/Component/Validator/Constraints/IbanValidator.php#L162

afurculita commented:
@rgeraads I have removed the 4 character length as I consider it too to be obsolete now. 
@fabpot I have removed the case check, as I consider it too unnecessary.

fabpot commented on src/Symfony/Component/Validator/Constraints/Iban.php:
```diff
@@ -23,18 +23,18 @@
+    const INVALID_CHARACTERS_ERROR = 2;
+    const CHECKSUM_FAILED_ERROR = 3;
+    const INVALID_FORMAT_ERROR = 4;
+    const NOT_SUPPORTED_COUNTRY_CODE_ERROR = 5;
```
You should revert the value changes as it would be a BC break. You should also keep the `TOO_SHORT_ERROR` to avoid BC breaks, and deprecate it so that we can remove it in 3.0.

fabpot replied:
The same goes for INVALID_CASE_ERROR

fabpot commented:
@afurculita Can you check the constants comment? As this is a bug fix, it should be merged into 2.3, but cannot contain BC breaks.

afurculita commented:
@fabpot The constant changes has been reverted and `TOO_SHORT_ERROR` and `INVALID_CASE_ERROR` are deprecated

fabpot commented on src/Symfony/Component/Validator/Constraints/Iban.php:
```diff
@@ -23,18 +23,22 @@
+    const NOT_SUPPORTED_COUNTRY_CODE_ERROR = 7;
 
     protected static $errorNames = array(
-        self::TOO_SHORT_ERROR => 'TOO_SHORT_ERROR',
```
should be kept as well for BC purposes.

afurculita replied:
Reverted this too

fabpot commented:
Thank you @afurculita.

SpacePossum commented:
thanks @afurculita, great to have this : D    :+1:
